### PR TITLE
feat: allow scopes for self signed jwt

### PR DESCRIPTION
**Pull request by arithmetic1728**
This PR allows self signed jwt to use scopes. 
AIP: https://google.aip.dev/auth/4111
googlers see: go/yoshi-self-signed-jwt-phase-2

In ServiceAccountCredentials, this PR now uses JwtCredentials instead of ServiceAccountJwtCredentials. As a result, this PR reverted the changes made to ServiceAccountJwtCredentials in #572 #642 so the ServiceAccountJwtCredentials is the same as what it was before phrase 1. 

The current behavior for ServiceAccountCredentials is:
```
if (hasScopes):
    if (useJwtAccessWithScope):
        // create a self signed JWT with "scope" set to the scope
    else:
        // call oauth token endpoint
else:
        // create a self signed JWT with modified uri as the audience
``` 

Follow up PRs:
(1) gax-java: https://github.com/googleapis/gax-java/pull/1420
(2) gapic-generator-java: it will be a very simple change, same as https://github.com/arithmetic1728/java-kms/pull/2/files. Since Gapic clients will always sets scopes, once self signed JWT is enabled in the future, it will always use self signed JWT with scope claim.

This PR has been tested with cloudkms: https://github.com/arithmetic1728/java-kms/pull/2/files

### CODE REVIEW

**Review by Neenu1995 — APPROVED** [review]

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 1]
```diff
@@ -950,14 +985,19 @@ public void getRequestMetadata(
   /** Provide the request metadata by putting an access JWT directly in the metadata. */
   @Override
   public Map<String, List<String>> getRequestMetadata(URI uri) throws IOException {
-    if (scopes.isEmpty() && defaultScopes.isEmpty() && uri == null) {
-      throw new IOException(
-          "Scopes and uri are not configured for service account. Either pass uri"
-              + " to getRequestMetadata to use self signed JWT, or specify the scopes"
-              + " by calling createScoped or passing scopes to constructor.");
+    if (createScopedRequired()) {
```
Comment (TimurSadykov) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:988]: i think we don't want to use this property here. It seems like not correct at the moment, because CreateScoped might not be actually required and Scopes could be empty at the same time. You can see functional tests examples as the reference. First couple 200 test cases will return true for createScopedRequired. The reason is that, as I understand, ESF will set scope based on audience. 

Therefore, I think that property is subject to change in the future.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java [hunk 2]
```diff
@@ -332,35 +331,17 @@ public boolean hasRequestMetadataOnly() {
     return true;
   }
 
-  /**
-   * Self signed JWT uses uri as audience, which should have the "https://{host}/" format. For
```
Comment (TimurSadykov) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java:336]: why this no longer needed?

**Review by TimurSadykov — CHANGES_REQUESTED** [review]

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 3]
```diff
@@ -492,7 +484,8 @@ static ServiceAccountCredentials fromPkcs8(
         serviceAccountUser,
         projectId,
         quotaProject,
-        DEFAULT_LIFETIME_IN_SECONDS);
+        DEFAULT_LIFETIME_IN_SECONDS,
+        false);
```
[the remark this answers is not in the document]
Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:488]: This feature is disabled by default because it doesn't work for all client libraries. Gapic clients will set this flag to `true` to enable the feature.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java (hunk 2 above)
Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java:336]: This was added by me in phrase 1, so I could create a ServiceAccountJwtAccessCredentials inside service account credentials to do the actual self signed jwt work. In this PR I switched to use JwtCredentials instead, so I reverted the changes made to ServiceAccountJwtAccessCredentials.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 4]
```diff
@@ -937,11 +949,34 @@ String createAssertionForIdToken(
...
+        .build();
+  }
+
   @Override
   public void getRequestMetadata(
       final URI uri, Executor executor, final RequestMetadataCallback callback) {
-    if (jwtCredentials != null && uri != null) {
-      jwtCredentials.getRequestMetadata(uri, executor, callback);
+    if (alwaysUseJwtAccess) {
+      // This will call getRequestMetadata(URI uri), which handles self signed JWT logic.
```
[the remark this answers is not in the document]
Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:978]: executor is for async, here we just do a blocking call to create a self signed jwt for better efficiency, similar as what ServiceAccountJwtAccessCredentials does:

https://github.com/googleapis/google-auth-library-java/blob/dfe118c261aadf137a3cf47a7acb9892c7a6db4d/oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java#L353-L358

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 1 above)
Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:988]: This is the original behavior of the library. I would like to keep the original behavior in this PR so it doesn't cause any problems.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 5]
```diff
@@ -698,7 +691,8 @@ public GoogleCredentials createScoped(
         serviceAccountUser,
         projectId,
         quotaProjectId,
-        lifetime);
+        lifetime,
+        alwaysUseJwtAccess);
```
[the remark this answers is not in the document]
Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:695]: The constructor has default visibility, I think it is only visible to this package. So it won't be a breaking change to the users.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 6]
```diff
@@ -728,7 +732,8 @@ public GoogleCredentials createDelegated(String user) {
         user,
         projectId,
         quotaProjectId,
-        lifetime);
+        lifetime,
+        alwaysUseJwtAccess);
```
[the remark this answers is not in the document]
Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:736]: I don't think so. See the comment above.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 7]
```diff
@@ -109,9 +110,9 @@
   private final Collection<String> defaultScopes;
   private final String quotaProjectId;
   private final int lifetime;
+  private final boolean alwaysUseJwtAccess;
```
[the remark this answers is not in the document]
Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: done

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 1 above)
Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: yep you are right, either scopes or uri work.

Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:988]: Nope. I would like to revert to the commit before phrase 1. In other words, always use scopes and never use uri.

https://github.com/googleapis/google-auth-library-java/blob/497d4e739cf83876b77d1eda8c9618968cf8165a/oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java#L414-L418

The integration test is not correct, it should be updated. It works because it uses self signed jwt with a modified url as audience (used the `https://<api>.googleapis.com/` part from the url). This is what I introduced in phrase 1. However this logic is wrong, url may be a regional endpoint or any url unrelated to audience, there is no way to derive the audience from uri. This is also the reason why I reverted the url modification logic in ServiceAccountJwtAccessCredentials in this PR. 
 
In other words, we should always scopes (never use uri) for service account credentials for Java. In this PR if `useJwtAccessWithScope` is `true`, I still included the `uri` logic because it won't be used by client libraries - GAPIC clients always set scopes we will always use scopes.

Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:987]: I updated the PR to only use scopes and never use uri, and updated the functional tests.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:977]: maybe add a comment

[the remark this answers is not in the document]
Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:977]: done

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 8]
```diff
@@ -109,9 +110,9 @@
   private final Collection<String> defaultScopes;
   private final String quotaProjectId;
   private final int lifetime;
+  private final boolean useJWTAccessWithScope;
```
Comment (TimurSadykov) [inline oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:113]: nit: this should be useJwtAccessWithScope. Sorry, I misstyped my suggestion :) per Jon, if used within name, JWT should be Jwt.

Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: yea, updated the name.

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java (hunk 1 above)
Comment (TimurSadykov) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java:989]: Thanks for explaining the change! However, the fact that those functional tests will fail with the change - means this is a breaking change, right? People could have used it this way already.

Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: added uri back

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java [hunk 9]
```diff
@@ -1387,7 +1461,11 @@ private void verifyJwtAccess(Map<String, List<String>> metadata) throws IOExcept
         JsonWebSignature.parse(GsonFactory.getDefaultInstance(), assertion);
     assertEquals(CLIENT_EMAIL, signature.getPayload().getIssuer());
     assertEquals(CLIENT_EMAIL, signature.getPayload().getSubject());
-    assertEquals(JWT_AUDIENCE, signature.getPayload().getAudience());
+    if (expectedScopeClaim != null) {
+      assertEquals(expectedScopeClaim, signature.getPayload().get("scope"));
```
Comment (TimurSadykov) [inline oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java:1465]: add check that aud is not present

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java [hunk 10]
```diff
@@ -1387,7 +1461,11 @@ private void verifyJwtAccess(Map<String, List<String>> metadata) throws IOExcept
...
+    } else {
+      assertEquals(JWT_AUDIENCE, signature.getPayload().getAudience());
```
Comment (TimurSadykov) [inline oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java:1467]: add check that scope is not present

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountJwtAccessCredentialsTest.java [hunk 11]
```diff
@@ -172,29 +171,6 @@ public void getAuthenticationType_returnsJwtAccess() throws IOException {
     assertEquals(credentials.getAuthenticationType(), "JWTAccess");
   }
 
-  @Test
```
Comment (TimurSadykov) [inline oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountJwtAccessCredentialsTest.java:175]: restore these tests for JWTCredential? Since you have moved the logic there

**Review by TimurSadykov — CHANGES_REQUESTED** [review]

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountJwtAccessCredentialsTest.java (hunk 11 above)
Comment (arithmetic1728) [reply oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountJwtAccessCredentialsTest.java:175]: These tests are moved to ServiceAccountCredentialsTest.

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java (hunk 10 above)
Comment (arithmetic1728) [reply oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java:1468]: done

File: oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java (hunk 9 above)
Comment (arithmetic1728) [reply oauth2_http/javatests/com/google/auth/oauth2/ServiceAccountCredentialsTest.java:1465]: done

File: oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java [hunk 12]
```diff
@@ -950,17 +1009,31 @@ public void getRequestMetadata(
...
+          "Scopes and uri are not configured for service account. Specify the scopes"
+              + " by calling createScoped or passing scopes to constructor or"
+              + " providing uri to getRequestMetadata.");
     }
-    if (jwtCredentials != null && uri != null) {
-      return jwtCredentials.getRequestMetadata(uri);
-    } else {
+
+    // If scopes are provided but we cannot use self signed JWT, then use scopes in the normal oauth
+    // way.
```
[the remark this answers is not in the document]
Comment (arithmetic1728) [reply oauth2_http/java/com/google/auth/oauth2/ServiceAccountCredentials.java]: done

**Review by TimurSadykov — APPROVED** [review]
